webapp/app/forms/steps/lotse/personal_data.py

```diff
@@ -91,15 +93,29 @@ def __init__(self, *args, **kwargs):
                             county.get('tax_offices')]
             self.bufa_nr.choices = choices
 
-        def validate_bundesland(form, field):
+        def validate_bundesland(form, field):            
             if form.steuernummer_exists.data == 'yes' or form.steuernummer_exists.data == 'no':
                 validators.InputRequired()(form, field)
             else:
                 validators.Optional()(form, field)
 
         def validate_steuernummer(form, field):
-            if form.steuernummer_exists.data == 'yes' and form.bundesland:
-                validators.InputRequired()(form, field)
+            if form.steuernummer_exists.data == 'yes':
+                bundesland_with_10_digits = {'BW', 'BE', 'HB', 'HH', 'NI', 'RP', 'SH'}
+                bundesland_with_11_digits = {'BY', 'BB', 'MV', 'NW', 'SL', 'SN', 'ST', 'TH'}
+                steuernummer_length = len(form.steuernummer.data);
+                
+                if form.bundesland:
+                    # Steuernummer is empty
+                    if not form.steuernummer.data or steuernummer_length == 0:                        
+                        raise validators.ValidationError(_l('form.lotse.steuernummer.input_required'))                
+                    elif form.bundesland.data in bundesland_with_10_digits and steuernummer_length < 10:
+                        raise validators.ValidationError(_l('form.lotse.steuernummer.input_required_with_10_digits'))                  
+                    elif form.bundesland.data in bundesland_with_11_digits and steuernummer_length < 11:
+                        raise validators.ValidationError(_l('form.lotse.steuernummer.input_required_with_11_digits'))            
+                    elif form.bundesland.data == 'HE' and steuernummer_length == 10:
+                        raise validators.ValidationError(_l('form.lotse.steuernummer.input_required_with_hessen_11_digits'))
+                        
```

Comment:
I am not sure if these changes should be part of this PR?

Reply:
Sorry for that